src/python/pants/engine/target.py

```diff
@@ -306,17 +307,6 @@ def __init__(
     def field_types(self) -> Tuple[Type[Field], ...]:
         return (*self.core_fields, *self.plugin_fields)
 
-    @union
-    @final
-    class PluginField:
```

Comment:
Rather than using `UnionMembership` for `Target` we began to break the dam with `def target_types` as an extension point in `register.py`. But I think that I would still prefer doubling down on union here (and there, perhaps) by:
* applying the `def PluginField` hack to give each subclass a unique type and avoid 1) the new extension point, 2) the new concept.
* giving `@union` an optional target/super class parameter that members should match

The reasoning for this is that there is effectively no end to the types of things that go into unions, and it feels arbitrary to move fields out (just as it did to move targets out). Folks cannot avoid learning about `@unions`: right now it might feel like they can, but only because folks aren't writing v2 plugins. To the extent that they have v1 targets to write bindings for, it's only because they actually wrote v1 plugins, and will need to port them to v2 and learn these concepts. No one will write _only_ a target or _only_ add a field: they will also need to learn about `@rules` and unions in order to actually consume targets/fields... so we're only adding concepts by moving them out.

Reply:
> No one will write only a target or only add a field: they will also need to learn about @rules and unions in order to actually consume targets/fields... so we're only adding concepts by moving them out.

Agreed with this. A new field on a pre-existing target type has zero use unless you write a plugin that consumes that new field.

Or, unless we add an option to `filter` to filter via field alias, e.g. all fields with `timeout`. Similar to filtering by tags.